### PR TITLE
Homedir files owners

### PR DESCRIPTION
**Pull request by jamescassell**
A possible alternative to #110 .

### CODE REVIEW

**Comment by mikerenfro** [comment]
Your find command for 020690 looks good so far. I haven't gotten the one for 020660 to work in my test cases yet, but I'm going to look at it a while longer. Thanks.

**Comment by mikerenfro** [comment]
I think I have a working 020660 find command based off of yours. Bear with me on the verification.

Test cases from my ansible user home directory:

```
[root@ansible ansible]# ls -alR
.:
total 20
drwxr-x---. 4 ansible ansible 155 Apr  4 08:43 .
drwxr-xr-x. 3 root    root     21 Mar 30 15:27 ..
-rw-------. 1 ansible ansible  18 Apr  3 14:04 .bash_history
-rw-r--r--. 1 ansible ansible  18 Aug  2  2017 .bash_logout
-rw-r--r--. 1 ansible ansible 193 Aug  2  2017 .bash_profile
-rw-r--r--. 1 ansible ansible 231 Aug  2  2017 .bashrc
-rw-r--r--. 1    2000 ansible 334 Sep 20  2017 .emacs
drwx------. 2 ansible ansible  57 Apr  2 17:17 .ssh
-rw-r--r--. 1 root    root      0 Apr  2 11:46 top-level-file
drwxr-xr-x. 2 root    root     46 Apr  2 14:03 top-level-folder

./.ssh:
total 12
drwx------. 2 ansible ansible   57 Apr  2 17:17 .
drwxr-x---. 4 ansible ansible  155 Apr  4 08:43 ..
-rw-------. 1 ansible ansible 1679 Mar 30 15:27 id_rsa
-rw-r--r--. 1 ansible ansible  421 Mar 30 15:27 id_rsa.pub
-rw-r--r--. 1 root    ansible  350 Apr  3 09:24 known_hosts

./top-level-folder:
total 0
drwxr-xr-x. 2 root    root     46 Apr  2 14:03 .
drwxr-x---. 4 ansible ansible 155 Apr  4 08:
-rw-r--r--. 1    2000 root      0 Apr  2 11:56 .lower-dotfile
-rw-r--r--. 1    2000 root      0 Apr  2 14:03 lower-file
```

Things to find and fix:

- `.emacs` should be owned by either root or ansible, not uid 2000 (020690)
- `top-level-folder/lower-file` and `top-level-folder/.lower-dotfile` should be owned by ansible (020660)
- `.ssh/known_hosts` should be owned by ansible (020660)
- `top-level-file` and `top-level-folder` should be owned by ansible (020660)

# 020660

Proposed find command for everything that's not a top-level dotfile (doesn't include a uid check yet):

```
find /home/ansible -mindepth 1 \
    \( -path "/home/ansible/.*/*" -o -path "/home/ansible/[^.]*" \
    -o \( ! -type f -path "/home/ansible/.*" \) \)
```

results in

```
/home/ansible/.ssh
/home/ansible/.ssh/id_rsa
/home/ansible/.ssh/id_rsa.pub
/home/ansible/.ssh/known_hosts
/home/ansible/top-level-file
/home/ansible/top-level-folder
/home/ansible/top-level-folder/.lower-dotfile
/home/ansible/top-level-folder/lower-file
```

which correctly ignored `.emacs` and `.bash*`, but caught `.lower-dotfile` and everything in the `.ssh` directory. And that looks like everything that's not a top-level dotfile.

Adding `-not -user 1000` to the find command results in

```
find /home/ansible -mindepth 1 \
    \( -path "/home/ansible/.*/*" -o -path "/home/ansible/[^.]*" \
    -o \( ! -type f -path "/home/ansible/.*" \) \) -not -user 1000
```

and

```
/home/ansible/.ssh/known_hosts
/home/ansible/top-level-file
/home/ansible/top-level-folder
/home/ansible/top-level-folder/.lower-dotfile
/home/ansible/top-level-folder/lower-file
```

which matches the files and folders that should be owned by uid 1000 to satisfy (my interpretation of) 020660.

# 020690

This one of yours worked out of the box, and I mostly just inverted the selection logic to find everything that wasn't a top-level dotfile for 020660.

```
find /home/ansible -mindepth 1 -type f \
    -path "/home/ansible/.*" -not -path "/home/ansible/.*/*" \
    -not \( -user 1000 -o -user root \)
```

results in

```
/home/ansible/.emacs
```

the only top-level dotfile not owned by root or uid 1000.

Thanks again for working this out.

**Comment by jamescassell** [comment]
for 020660, you need to add ` -print` to the `{{ find_command_base }}` in order for it to work correctly.  If you run the role with my PR as-is, does it give unexpected results?  (such as running it with `--tags=RHEL-07-020660`)

**Comment by mikerenfro** [comment]
Didn't get that far into running the actual playbook, I just used the find command at to the command prompt, and I assumed `-print` would have been inherent. But I was apparently wrong, for some reason.

Without `-print`:

```
[root@ansible ansible]# find "/home/ansible" -mindepth 1 \( -path "/home/ansible/.*" -not -path "/home/ansible/.*/*" -type f \) -prune -o -not -user 1000
/home/ansible/.bash_logout
/home/ansible/.bash_profile
/home/ansible/.bashrc
/home/ansible/.emacs
/home/ansible/.ssh/known_hosts
/home/ansible/top-level-file
/home/ansible/top-level-folder
/home/ansible/top-level-folder/.lower-dotfile
/home/ansible/top-level-folder/lower-file
/home/ansible/.bash_history
```

with `-print`:

```
[root@ansible ansible]# find "/home/ansible" -mindepth 1 \( -path "/home/ansible/.*" -not -path "/home/ansible/.*/*" -type f \) -prune -o -not -user 1000 -print
/home/ansible/.ssh/known_hosts
/home/ansible/top-level-file
/home/ansible/top-level-folder
/home/ansible/top-level-folder/.lower-dotfile
/home/ansible/top-level-folder/lower-file
```

so your 020660 was fine all along. My mistake.

**Comment by jamescassell** [comment]
Seems that if a `-print` -or `-exec` type of action is not explicit somewhere in the command, it assumes a '-print' at the end and assumes parenthesis around the entire expression before the -print

File: tasks/fix-cat2.yml [hunk 1]
```diff
@@ -823,8 +823,8 @@
             this_item: "{{ item.item }}"
   vars:
       find_command_base: 'find "{{ this_item.dir }}" -mindepth 1
-             ( -path "{{ this_item.dir }}/.*" -not -path "{{ this_item.dir }}/.*/*" -type f )
-             -prune -o -not -user {{ this_item.uid }}'
+             ( -path "{{ this_item.dir }}/.*" -not -path "{{ this_item.dir }}/.*/*" -type f ) -o
```
Comment (jamescassell) [inline tasks/fix-cat2.yml]: the parens here are technically unnecessary, but I put them to be explicit...

**Comment by jamescassell** [comment]
added RHEL-07-020670 and RHEL-07-020700.

**Pull request merged**